i6_models/parts/mha.py

```diff
@@ -0,0 +1,100 @@
+
+
+class MultiheadAttentionV1(torch.nn.Module):
+    """
+    Native Multihead Attention implementation based on 'Attention Is All You Need'
+    """
+
+    def __init__(self, cfg: MultiheadAttentionV1Config):
+        super().__init__()
+        self.cfg = cfg
+        self.num_att_heads = cfg.num_att_heads
+        self.input_dim = cfg.input_dim
+        self.dim_heads = self.input_dim // self.num_att_heads
+        self.layernorm = torch.nn.LayerNorm(cfg.input_dim)
+
+        self.out_proj = torch.nn.Linear(in_features=cfg.input_dim, out_features=cfg.input_dim, bias=True)
+        self.in_proj = torch.nn.Linear(in_features=cfg.input_dim, out_features=3 * cfg.input_dim, bias=True)
+
+        self.norm = math.sqrt(float(self.input_dim / self.num_att_heads))
+        self.softmax = torch.nn.Softmax(-1)
+        self.dropout = torch.nn.Dropout(cfg.att_weights_dropout)
+
+    def forward(self, query: torch.Tensor, key: torch.Tensor, value: torch.Tensor, key_padding_mask: torch.Tensor):
+        """
+        Computes the forward pass of the MultiheadAttentionV1 module.
+
+        Attributes:
+            query (torch.Tensor): The input query tensor of shape (B, T, F).
+            key (torch.Tensor): The input key tensor of shape (B, T, F).
+            value (torch.Tensor): The input value tensor of shape (B, T, F).
+            key_padding_mask (torch.Tensor): The key padding mask tensor of shape (batch_dim, num_tokens).
+
+        Note:
+            - Only self attention is supported for now.
+            - The key_padding_mask tensor is expected to have the shape (batch_dim, num_tokens).
+        """
+
+        assert query is value is key, "only supports self attention for now"
+
+        batch_dim, num_tokens, embed_dim = query.shape
+        x = self.in_proj(query)
+
+        hidden_dim = query.size(-1)
+        query, key, value = x.unflatten(-1, (3, hidden_dim)).unsqueeze(0).transpose(0, -2).squeeze(-2).contiguous()
+
+        query = query.view(batch_dim, -1, self.num_att_heads, self.dim_heads)  # [B, T, D//H, D']
+        key = key.view(batch_dim, -1, self.num_att_heads, self.dim_heads)  # [B, T, D//H, D']
+        value = value.view(batch_dim, -1, self.num_att_heads, self.dim_heads)  # [B, T, D//H, D']
+
+        query = torch.transpose(query, 1, 2)  # [B, D//H, T, D']
+        key = torch.transpose(key, 1, 2)  # [B, D//H, T, D']
+        value = torch.transpose(value, 1, 2)  # [B, D//H, T, D']
+
+        key = torch.transpose(key, -2, -1)  # [B, D//H, D', T]
+
+        dot = torch.matmul(query, key)  # [B, D//H, T, T]
+        dot = dot / self.norm
+
+        if key_padding_mask is not None:
+            key_padding_mask = key_padding_mask.view(batch_dim, 1, 1, key_padding_mask.size(1))
+            dot = dot.masked_fill(key_padding_mask, -float("inf"))
+
+        alpha = self.softmax(dot)  # [B, D//H, T, T]
+        alpha = self.dropout(alpha)
```

Comment:
Is it intended that the dropout acts on the alphas (i.e. the attention weight?)? 

This would make them non-normalized. I would have dropped out the `dot` instead, to keep normalization.. but if it is _always_ done this way, who am I to fight.. (But still, would this change in weight not have to be compensated in recognition?)

Also the docstring for the parameter reads
```
dropout: multi-headed self attention output dropout
```
so I would expect that the dropout is applied to the output of this module (i.e. after `self.out_proj`)

Reply:
So if i am not lost in the torch code they also do it this way. I do see your point though, not sure if it makes sense to dropout the weights but rather the dot. 

https://github.com/pytorch/pytorch/blob/6e43897912d149d7dad676f496c608fe32a31978/torch/nn/functional.py#L6212

Reply:
> they also do it this way. 

yes. If the experts say this is how its always done, then I am fine to also have it like this. 

Then I would only suggest to change the documentation to e.g.
```
dropout: attention weight dropout probability
```